File: java/530.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int ans = Integer.MAX_VALUE;
    TreeNode prev = null; 
    public int getMinimumDifference(TreeNode root) {
        // System.out.println(ans);
        dfs(root);
        return (int)ans;
    }
    
    public void dfs(TreeNode root){
        if(root != null){
            dfs(root.left);
           if(prev != null){
               ans = Math.min(ans,Math.abs(root.val - prev.val));
           }
            prev = root;
            dfs(root.right);
        }
    }
}

File: python/530.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def getMinimumDifference(self, root):

        """
        :type root: TreeNode
        :rtype: int
        """
        # self.ans = float("inf")
        # self.val = None
        # def inorder(root):
        #     if root == None:
        #         return
        #     inorder(root.left)
        #     if self.val is not None:
        #         self.ans = min(self.ans, abs(root.val - self.val))
        #     self.val = root.val
        #     inorder(root.right)
        # inorder(root)
        # return self.ans
        self.mini = 2 ** 31
        self.pre = -1

        def inorder(root):
            if root == None:
                return self.mini
            inorder(root.left)

            if self.pre != -1:
                self.mini = min(self.mini,abs(root.val - self.pre))
            self.pre = root.val
            inorder(root.right)
            return self.mini
        inorder(root)
        return self.mini
